src/Scenario.cpp

```
@@ -0,0 +1,382 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "rmf_performance_tests/Scenario.hpp"
+
+void rmf_performance_tests::scenario::parse_scenario(std::string scenario_file,
+  Scenario& scenario)
+{
+  if (scenario_file.rfind(".yaml") == std::string::npos)
+  {
+    scenario_file.append(".yaml");
+  }
+
+  YAML::Node scenario_config;
+
+  try
+  {
+    scenario_config = YAML::LoadFile(std::string(
+          TEST_SCENARIO_DIR) + scenario_file);
+  }
+  catch (YAML::BadFile& e)
+  {
+    throw std::runtime_error(
+            "Failed to load scenario file [" + scenario_file + "]");
+  }
+
+  if (scenario_config["samples"])
+  {
+    scenario.samples = scenario_config["samples"].as<std::size_t>(100);
+  }
+  else
+  {
+    std::cout <<
+      "Scenario file is missing the [samples] key. Using default value [100]" <<
+      std::endl;
+  }
+
+  const YAML::Node robots = scenario_config["robots"];
+
+  for (auto iter = robots.begin(); iter != robots.end(); ++iter)
+  {
+    const auto& name = iter->first.as<std::string>();
+    const auto& robot = iter->second;
+
+    double linear_velocity, linear_acceleration, angular_velocity,
+      angular_acceleration;
+
+    const auto& limits = robot["limits"];
+    if (limits)
+    {
+      const auto& linear = limits["linear"];
+      const auto& angular = limits["angular"];
+
+      if (linear)
+      {
+        const auto& velocity = linear["velocity"];
+        const auto& acceleration = linear["acceleration"];
+
+        if (velocity)
+        {
+          linear_velocity = velocity.as<double>();
+        }
+        else
+        {
+          std::cout << "Robot [" << name <<
```

This is a note for all parsing errors where items are missing:

We should not skip over mandatory missing fields. Instead we should throw an exception and force the test to terminate. It would be deceptive for users if they get a performance result for a scenario that they didn't really ask for.

We can take advantage of `YAML::ParserException` which helps users see which exact line is wrong. For example, here we would do this:

```
throw YAML::ParserException(
  velocity.Mark(),
  "Robot [" + name + "] is missing key [" + key_velocity + "]");
```